File: javascript/adapters/static_server.js
Faye.StaticServer = Faye.Class({
  initialize: function(directory, pathRegex) {
    this._directory = directory;
    this._pathRegex = pathRegex;
    this._pathMap   = {};
    this._index     = {};
  },
  
  map: function(requestPath, filename) {
    this._pathMap[requestPath] = filename;
  },
  
  test: function(pathname) {
    return this._pathRegex.test(pathname);
  },
  
  serve: function(pathname, request, response) {
    pathname = path.basename(pathname);
    pathname = this._pathMap[pathname] || pathname;
    
    this._index[pathname] = this._index[pathname] || {};
    
    var cache    = this._index[pathname],
        fullpath = path.join(this._directory, pathname);
    
    cache.content = cache.content || fs.readFileSync(fullpath);
    cache.digest  = cache.digest  || crypto.createHash('sha1').update(cache.content).digest('hex');
    cache.mtime   = cache.mtime   || fs.statSync(fullpath).mtime;
    
    var headers = Faye.extend({}, Faye.NodeAdapter.prototype.TYPE_SCRIPT),
        ims     = request.headers['if-modified-since'];
    
    headers['ETag'] = cache.digest;
    headers['Last-Modified'] = cache.mtime.toGMTString();
    
    if (request.headers['if-none-match'] === cache.digest) {
      headers['Content-Length'] = '0';
      response.writeHead(304, headers);
      response.end();
    }
    else if (ims && cache.mtime <= new Date(ims)) {
      headers['Content-Length'] = '0';
      response.writeHead(304, headers);
      response.end();
    }
    else {
      headers['Content-Length'] = cache.content.length;
      response.writeHead(200, headers);
      response.write(cache.content);
      response.end();
    }
  }
});

File: lib/faye/adapters/static_server.rb
module Faye
  class StaticServer

    def initialize(directory, path_regex)
      @directory  = directory
      @path_regex = path_regex
      @path_map   = {}
      @index      = {}
    end

    def map(request_path, filename)
      @path_map[request_path] = filename
    end

    def =~(pathname)
      @path_regex =~ pathname
    end

    def call(env)
      filename = File.basename(env['PATH_INFO'])
      filename = @path_map[filename] || filename

      cache = @index[filename] ||= {}
      fullpath = File.join(@directory, filename)

      begin
        cache[:content] ||= File.read(fullpath)
        cache[:digest]  ||= Digest::SHA1.hexdigest(cache[:content])
        cache[:mtime]   ||= File.mtime(fullpath)
      rescue
        return [404, {}, []]
      end

      type = /\.js$/ =~ fullpath ? RackAdapter::TYPE_SCRIPT : RackAdapter::TYPE_JSON
      ims  = env['HTTP_IF_MODIFIED_SINCE']

      no_content_length = env[RackAdapter::HTTP_X_NO_CONTENT_LENGTH]

      headers = {
        'ETag'          => cache[:digest],
        'Last-Modified' => cache[:mtime].httpdate
      }

      if env['HTTP_IF_NONE_MATCH'] == cache[:digest]
        [304, headers, ['']]
      elsif ims and cache[:mtime] <= Time.httpdate(ims)
        [304, headers, ['']]
      else
        headers['Content-Length'] = cache[:content].bytesize.to_s unless no_content_length
        headers.update(type)
        [200, headers, [cache[:content]]]
      end
    end

  end
end
